### DISCUSSION TITLE
About Vertical Expansion and Resource Allocation 关于纵向扩展和资源分配

### DISCUSSION THREAD

**Discussion by IeohMingChan** [milvus-io/milvus#34128] [Q&A and General discussion]
We are planning to use Milvus to establish a vector database for project business, and the number of vectors stored in the database will gradually increase over time, so the demand for the k8s cluster will also gradually increase. To improve the resource utilization of the k8s cluster, it is planned to apply for slightly fewer resources, especially the number of CPU cores, in the early stages of the project, and gradually expand resources according to business needs in the future. In the official documentation, only the content of milvus horizontal expansion appears, but according to the resource estimation of Milvus Sizing Tool, as the number of vectors increases, the worker nodes also need to be vertically expanded. Now I have the following question:
1. Will vertical expansion of worker nodes during the operation of milvus help affect the operation of milvus? Will the data already written be lost, especially? During this period, project business can be suspended and data reading and writing can be suspended.
2. Under limited resources, should we prioritize increasing the number of worker nodes or increasing the resource quota for individual worker nodes to improve read and write performance?
Thank you very much for your answers!

我们正计划使用milvus建立项目业务的向量数据库，数据库保存向量数将随时间增长而逐渐增加，因此对k8s集群的需求也逐渐增大。为提高k8s集群的资源利用率，在项目初期计划申请稍微少一些的资源，尤其是CPU核心数，后续再根据业务需求逐渐扩展资源。在官方文档中，只出现了milvus横向扩展的内容，但根据Milvus Sizing Tool的资源估计，随着向量数的增加，工作节点也需要进行纵向扩展。现在我有如下问题：
1. 请问在milvus helm运行过程中对工作节点进行纵向扩展是否会影响milvus的运行？尤其是已经写入的数据是否会丢失？在此期间可以暂停项目业务，暂停数据读写。
2. 请问在资源有限的条件下，应该优先增加工作节点数量还是增加单个工作节点的资源配额，以提高读写性能？
万分感谢大家的回答！

**Comment by yhmo** [comment]
When you insert some data by insert():
1. The data is received by proxy node, proxy node splits the data into pieces by shards_num(which is defined in collection schema, default is 1), and sends the pieces(as messages) into Pulsar/Kafka. The insert() returns success to the client once the data is persisted into Pulsar/Kafka.
2. For each growing segment, two "data channels" are created, one is between querynode and Pulsar/Kafka, the other is between datanode and Pulsar/Kafka.
3. Querynode and datanode subscribe messages asynchronously from Pulsar/Kafka, the data pieces are consumed by querynode and datanode.

Pulsar/Kafka is like a WAL(write-ahead-log) component for Milvus. If querynode or datanode crashes, they will reconsume the data from a certain timepoint. So, data is ensure to be safe.

**Comment by yhmo** [comment]
When you call collection.load(), there is a parameter replica_number, default value is 1.
If you set replica_numer=2, that means the collection has two copies in memory.

Let's say you have 2 query nodes. A collection has 2 segments.

- If you set replica_num =1, the 2 segments will be averagely assigned to 2 query nodes, each node takes 1 segment. Now you want to  reallocate more CPU to a query node, you need to restart this query node. Since there is only one segment is in memory, the query service will be broken, until the new query node back to online.

- if you set replica_num=2, the 2 segments have two copies in memory of the 2 query nodes. Based on the internal balance logic, Milvus will assign one copy to node_1, another copy to node_2. Now you want to  reallocate more CPU to a query node, you need to restart this query node. Since there still is a copy in another query node, the query service will not be broken.

**Reply by IeohMingChan** [reply]
According to your answer, if I stop all query services first, then vertically expand the query nodes, and then restore the query services, it will not have any negative impact on milvus?

**Reply by yhmo** [reply]
No impact to data, but will stop the query service of your business.

**Comment by yhmo** [comment]
Typically, 4 ~ 8 CPU cores per query node and 2 CPU cores per data node is ok. No need to assign too many CPU cores to a query node.
.

**Reply by IeohMingChan** [reply]
As expected, the number of vectors we store will gradually increase to a maximum value of 19.5 million. According to Milvus Sizing Tool's estimate, storing 19.5M vectors requires 4 8-core 16GB index nodes and 4 8-core 32GB query nodes. At present, our k8s cluster only allows the deployment of one index node and one query node of the above specifications with remaining resources. At the current initial stage, should I directly deploy a single large index node and query, or split it into two or more query nodes and index nodes? Do you have a recommended allocation plan of memory and CPU? Thank you.

**Reply by yhmo** [reply]
Two nodes is better than one node.
No exact "best plan", just try. 16 cores one node is ok, 8 cores two nodes are also work.

**Reply by IeohMingChan** [reply]
I got it, thanks!

**Comment by xiaofan-luan** [comment]
1. we recommend to use K8s operator since it support rolling upgrade
2. scale up or scale out based on how much data you have. For any use case less than 10M, scale up is recommend.
